web/src/reactions.ts

```diff
@@ -36,18 +51,32 @@ function get_message(message_id) {
     return message;
 }
 
-function create_reaction(message_id, reaction_info) {
+export type RawReaction = {
+    emoji_name: string;
+    reaction_type: string;
+    emoji_code: string;
+    user_id: number;
+};
+
+function create_reaction(message_id: number, rendering_details: EmojiRenderingDetails): ReactionEvent {
+    assert(page_params.user_id !== undefined);
     return {
         message_id,
         user_id: page_params.user_id,
-        local_id: get_local_reaction_id(reaction_info),
-        reaction_type: reaction_info.reaction_type,
-        emoji_name: reaction_info.emoji_name,
-        emoji_code: reaction_info.emoji_code,
+        local_id: get_local_reaction_id(rendering_details),
+        reaction_type: rendering_details.reaction_type,
+        emoji_name: rendering_details.emoji_name,
+        // TODO(evy): deal with this number | string situation
+        // the emoji_code can be number because `active_realm_emojis`
+        // and the `RealmEmoji` type.
```

Comment:
@andersk looking for guidance here -- essentially sometimes it can only be a string (and that's what `MessageCleanReaction` was set up as) but because of `active_realm_emojis and the `RealmEmoji` type, here it wants to sometimes be a number. It seems like it would be annoying to allow `MessageCleanReaction` to have numbers, and also annoying to force it to be a string. Do you have advice on how to proceed?

Reply:
Fix `EmojiRenderingDetails` and `ReactionEvent` to model reality: `{…} & ({reaction_type: "unicode_emoji"; emoji_code: string} | {reaction_type: "realm_emoji" | "zulip_extra_emoji"; emoji_code: number})`.

Reply:
I tried that but it got pretty messy to appease typescript -- it seems like it doesn't understand that the constraint carries over between types? Maybe you can find something cleaner than what I tried. I just pushed my attempt. Also happy to try pairing on it